src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -3262,6 +3262,293 @@ void C2_MacroAssembler::stringL_indexof_char(Register str1, Register cnt1, Regis
+    break;
+  default:
+    ShouldNotReachHere();
+  }
+}
+
+void C2_MacroAssembler::arrays_hashcode_elvload(XMMRegister dst, Address src, BasicType eltype) {
+  load_vector(dst, src, arrays_hashcode_elsize(eltype) * 8);
+}
+
+void C2_MacroAssembler::arrays_hashcode_elvload(XMMRegister dst, AddressLiteral src, BasicType eltype) {
+  load_vector(dst, src, arrays_hashcode_elsize(eltype) * 8);
+}
+
+void C2_MacroAssembler::arrays_hashcode_elvcast(XMMRegister dst, BasicType eltype) {
+  switch (eltype) {
+  case T_BYTE:
+    vector_unsigned_cast(dst, dst, Assembler::AVX_256bit, T_BYTE, T_INT);
+    break;
+  case T_SHORT:
+  case T_CHAR:
+    vector_unsigned_cast(dst, dst, Assembler::AVX_256bit, T_SHORT, T_INT);
+    break;
+  case T_INT:
+    // do nothing
+    break;
+  default:
+    ShouldNotReachHere();
+  }
+}
+
+void C2_MacroAssembler::arrays_hashcode(Register ary1, Register cnt1, Register result,
+                                        Register index, Register coef, Register tmp, XMMRegister vnext,
+                                        XMMRegister vcoef0, XMMRegister vcoef1, XMMRegister vcoef2, XMMRegister vcoef3,
+                                        XMMRegister vresult0, XMMRegister vresult1, XMMRegister vresult2, XMMRegister vresult3,
+                                        XMMRegister vtmp0, XMMRegister vtmp1, XMMRegister vtmp2, XMMRegister vtmp3,
+                                        int mode) {
+  ShortBranchVerifier sbv(this);
+  assert(UseAVX >= 2, "AVX2 intrinsics are required");
+  assert_different_registers(ary1, cnt1, result, index, coef, tmp);
+  assert_different_registers(vnext, vcoef0, vcoef1, vcoef2, vcoef3, vresult0, vresult1, vresult2, vresult3, vtmp0, vtmp1, vtmp2, vtmp3);
+
+  Label SHORT_UNROLLED_LOOP_BEGIN, SHORT_SCALAR_LOOP_BEGIN,
+        LONG, LONG_INIT, LONG_SCALAR_LOOP_BEGIN, LONG_SCALAR_LOOP_END, LONG_VECTOR_LOOP_BEGIN, LONG_VECTOR_LOOP_END,
+        END;
+
+  // For "renaming" for readibility of the code
+  Register bound;
+  XMMRegister vcoef[] = { vcoef0, vcoef1, vcoef2, vcoef3 },
+              vresult[] = { vresult0, vresult1, vresult2, vresult3 },
+              vtmp[] = { vtmp0, vtmp1, vtmp2, vtmp3 };
+
+  const VectorizedHashCodeNode::HashMode hashMode = (VectorizedHashCodeNode::HashMode)mode;
+  const BasicType eltype = VectorizedHashCodeNode::adr_basic_type(hashMode);
+  const bool is_string_hashcode = hashMode == VectorizedHashCodeNode::LATIN1 || hashMode == VectorizedHashCodeNode::UTF16;
+  const bool is_unsigned = is_string_hashcode || hashMode == VectorizedHashCodeNode::CHAR;
+  const int elsize = arrays_hashcode_elsize(eltype);
+  // vector loops does not currently support short and signed byte arrays for now, need to figure out how to properly sign-extend when loading into vectors
+  const bool generate_vectorized_loop = is_unsigned || hashMode == VectorizedHashCodeNode::INT;
+
+  // int result = 0|1;
+  movl(result, is_string_hashcode ? 0 : 1);
+
+  // if (cnt1 == 0) {
```

Comment:
You may want to reorder the execution of the loops, a short array suffers more from processing than a big array, so you should have minimum extra hops for those. For example, I think this could be:

    if (cnt1 >= 4) {
        if (cnt1 >= 16) {
            UNROLLED VECTOR LOOP
            SINGLE VECTOR LOOP
        }
        UNROLLED SCALAR LOOP
    }
    SINGLE SCALAR LOOP

The thresholds are arbitrary and need to be measured carefully.

Reply:
Fixed